doc/SAI-Proposal-Prefix-Compression.md

```diff
@@ -0,0 +1,308 @@
+entry_v4_1.prefix.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+entry_v4_1.prefix.addr.ipv4 = "1.1.1.1";
+entry_v4_1.prefix_mask.ipv4 = "255.255.255.0";
+sai_entry_list[0].id = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_META;
+sai_entry_list[0].value.u32 = 2;
+
+// IPV4 Second Entry
+sai_prefix_compression_entry_t entry_v4_1;
+entry_v4_1.switch_id = switch_id;
+entry_v4_1.prefix_table_id = src_prefix_compression_table_id;
+entry_v4_1.prefix.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+entry_v4_1.prefix.addr.ipv4 = "2.2.2.1";
+entry_v4_1.prefix_mask.ipv4 = "255.255.255.0";
+sai_entries_attribute_list[0].id = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_META;
+sai_entries_attribute_list[0].value.u32 = 800;
+
+// IPV6 Entry
+sai_prefix_compression_entry_t entry_v6_1;
+entry_v6_1.switch_id = switch_id;
+entry_v6_1.prefix_table_id = src_prefix_compression_table_id;
+entry_v6_1.prefix.addr_family = SAI_IP_ADDR_FAMILY_IPV6;
+entry_v6_1.prefix.addr.ipv4 ="2001:1::4";
+entry_v6_1.prefix_mask = "ffff:ffff:ffff:ffff:ffff:ffff:ffff:fffc";
+sai_entries_attribute_list[1].id = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_META;
+sai_entries_attribute_list[1].value.u32 = 4;
+
+entries_list = [entry_v4_2, entry_v6_1];
+entries_list_size = 2;
+sai_status_t bulk_status;
+
+// Adding a single entry to a table
+sai_attr_list_count = 1;
+sai_create_prefix_compression_entry_fn(
+    entry_v4_1,
+    sai_attr_list,
+    sai_attr_list_count);
+
+// Adding using bulk add
+attr_entries_count = 2;
+sai_bulk_create_prefix_compression_entry_fn(
+    entries_list_size,
+    entries_list,
+    attr_entries_count,
+    sai_entries_attribute_list,
+    SAI_BULK_OP_ERROR_MODE_STOP_ON_ERROR,
+    bulk_status);
+
+// Example of using Prefix Compression with ACL
+// Create ACL table
+sai_table_attr_list[0].id = SAI_ACL_TABLE_ATTR_SRC_PREFIX_COMPRESSION_TABLE;
+sai_table_attr_list[0].value.oid = src_prefix_compression_table_id;
+sai_table_attr_count = 1;
+
+sai_create_acl_table_fn(
+    &acl_table_id,
+    switch_id,
+    sai_table_attr_count,
+    sai_table_attr_list);
+
+sai_entry_attr_list[0].id = SAI_ACL_ENTRY_ATTR_FIELD_SRC_PREFIX_META;
+sai_attr_list[1].value.aclfield.enable = true;
+sai_attr_list[1].value.aclfield.mask = 0xFFFFFF;
+sai_attr_list[1].value.aclfield.data = 800;
+sai_entry_attr_count = 1;
+
+// Create an Entry to match on META data
+sai_create_acl_entry_fn(
+    &acl_entry_id,
+    switch_id,
+    sai_entry_attr_count,
+    sai_entry_attr_list);
+
+// Remove ACL entry
+sai_create_acl_entry_fn(
+    acl_entry_id);
+
+// Remove ACL Table
+sai_create_acl_entry_fn(
+    acl_table_id);
+
+// Remove Single Entry
+sai_remove_prefix_compression_entry_fn(
+    entry_v4_1);
+
+// Remove Bulk Entries
+sai_bulk_remove_prefix_compression_entry_fn(
+    entries_list_size,
+    entries_list,
+    SAI_BULK_OP_ERROR_MODE_STOP_ON_ERROR,
+    bulk_status);
+
+// Remove Prefix Compression Table
+sai_remove_prefix_compression_table_fn(
+    src_prefix_compression_table_id);
+
+```
```

Comment:
Please consider adding the workflow to identify the platforms of the below types using sai\_object\_type\_get\_availability() and using filters on SAI\_PREFIX\_COMPRESSION\_TYPE:

1.  A single compression table of type SAI\_PREFIX\_COMPRESSION\_TYPE\_BOTH, 
2.  A compression table or type SAI\_PREFIX\_COMPRESSION\_TYPE\_SRC and another SAI\_PREFIX\_COMPRESSION\_TYPE\_DST